File: graph/2361.Minimum_Costs_Using_the_Train_Line.cpp
#include <vector>
#include <queue>
#include <utility>
#include <climits>

using namespace std;

class Solution {
public:
    vector<long long> minimumCosts(vector<int>& regular, vector<int>& express, int expressCost) {
        
        int num_of_stops = regular.size() + 1;

        vector<pair<int,int>> adj_graph[num_of_stops * 2];
        for (int i = 0; i < regular.size(); i++) {
            // auto p = make_pair(i+1, regular[i]);
            adj_graph[i].push_back(make_pair(i+1, regular[i]));
            adj_graph[i].push_back(make_pair(i + num_of_stops, expressCost));
        }
        for (int i = 0; i < express.size(); i++) {
            adj_graph[i + num_of_stops].push_back(make_pair(i + 1 + num_of_stops, express[i]));
            adj_graph[i + num_of_stops].push_back(make_pair(i, 0));
        }

        int dist_to[num_of_stops * 2];
        dist_to[0] = 0;
        for (int i = 1; i < num_of_stops * 2; i++) {
            dist_to[i] = INT_MAX;
        }
        priority_queue<pair<int, int>> pq;
        pq.push(make_pair(0, 0));
        while (!pq.empty()) {
            auto p = pq.top();
            pq.pop();
            int distance = p.first;
            int from = p.second;
            for (auto & p2: adj_graph[from]) {
                int to = p2.first;
                int cost = p2.second;
                if (dist_to[from] + cost < dist_to[to]) {
                    dist_to[to] = dist_to[from] + cost;
                    pq.push(make_pair(dist_to[to], to));
                }
            }
        }

        vector<long long> res;
        for (int i = 0; i < num_of_stops; i++) {
            res.push_back(min(dist_to[i], dist_to[i + num_of_stops]));
        }
        res.erase(res.begin());
        return res;
    }
};

File: graph/2361.Minimum_Costs_Using_the_Train_Line.py
# Dijkstra
class Solution:
    def minimumCosts(self, regular: List[int], express: List[int], expressCost: int) -> List[int]:

        num_of_stops = len(regular) + 1

        V = (num_of_stops) * 2
        graph = [[] for _ in range(V)]
        for i in range(len(regular)):
            graph[i].append((i+1, regular[i]))  # e.g. 0 -> (1, 1)
            graph[i].append((i + num_of_stops, expressCost))
        for i in range(len(express)):
            express_idx = i + num_of_stops
            graph[express_idx].append((express_idx + 1, express[i]))  # e.g. 3(express) -> (4(express), 10)
            graph[express_idx].append((i, 0))
        
        dist_to = [float('inf') for _ in range(V)]
        dist_to[0] = 0
        pq = []
        pq.append((0, 0))
        while pq:
            distance, node = heapq.heappop(pq)
            for neighbor, w in graph[node]:
                if distance + w < dist_to[neighbor]:
                    dist_to[neighbor] = distance + w
                    heapq.heappush(pq, (dist_to[neighbor], neighbor))

        # decode
        res = []
        for i in range(num_of_stops):
            res.append(min(dist_to[i], dist_to[i + num_of_stops]))
        res.pop(0)  # remove min distance to stop #0 (always 0)

        return res
